nemo_aligner/data/nlp/datasets.py

```diff
@@ -403,10 +432,153 @@ def __getitem__(self, idx):
+        def combine_keys(key):
+            return [item[key] for item in batch]
+
+        lengths = combine_keys("lengths")
+        rewards = combine_keys("reward")
+        seq_boundaries = combine_keys("seq_boundaries")
+
+        input_ids = [
+            np.concatenate(
+                [
+                    item["input_ids"][item["seq_boundaries"][i] : item["seq_boundaries"][i + 1] - 1]
+                    for i in range(len(item["seq_boundaries"]) - 1)
+                ]
+            )
+            for item in batch
+        ]
+        labels = [
+            np.concatenate(
+                [
+                    item["labels"][item["seq_boundaries"][i] + 1 : item["seq_boundaries"][i + 1]]
+                    for i in range(len(item["seq_boundaries"]) - 1)
+                ]
+            )
+            for item in batch
+        ]
+
+        if pad_length_to_multiple_of:
+            max_seq_len = torch.tensor(input_ids.shape[1], device=torch.cuda.current_device())
+            torch.distributed.all_reduce(
+                max_seq_len, op=torch.distributed.ReduceOp.MAX, group=parallel_state.get_data_parallel_group()
+            )
+            max_length = math.ceil(max_seq_len / pad_length_to_multiple_of) * pad_length_to_multiple_of
+        else:
+            # pad to the nearest multiple of 16 for FP8 training
+            # for many datasets in practice, all packed sequence lengths are very close to the
+            # target length (2048, 4096, 8192), so there is very minimal padding
+            max_length = max(len(l) for l in input_ids)
+            max_length = min(self.seq_length, self._ceil_to_nearest(max_length, 16))
+        assert max_length <= self.seq_length
+
+        position_ids: List[List[int]] = []
+        cu_seqlens: List[List[int]] = []
+        for item in batch:
+            position_ids.append([])
+            cu_seqlens.append([0])
+            seqlens = np.array(item["seq_boundaries"][1:]) - np.array(item["seq_boundaries"][:-1])
+            for l in seqlens:
+                position_ids[-1].extend(list(range(l - 1)))  ## l - 1 to exclude labels
+                cu_seqlens[-1].append(cu_seqlens[-1][-1] + l - 1)
+            # set last seq to the max seq len because rope and attn kernels expect no padding
+            cu_seqlens[-1][-1] = max_length
+
+        assert len(input_ids[0]) == len(
+            position_ids[0]
+        ), "Dataset problem: input_ids and position_ids lengths don't match"
+
+        input_ids = self._collate_item(input_ids, max_length=max_length, pad_id=self.tokenizer.eos_id)
+        labels = self._collate_item(labels, max_length=max_length, pad_id=-100)
+        position_ids = self._collate_item(position_ids, max_length=max_length, pad_id=0)
+
+        max_num_sequences = max(len(l) for l in lengths)
+        lengths = self._collate_item(lengths, max_length=max_num_sequences, pad_id=0)
+        rewards = self._collate_item(rewards, max_length=max_num_sequences, pad_id=-1000)
```

Comment:
Can we define a global or a class property like:

```
class DPOPackedDataset(...):

    REWARDS_PAD_ID = -1000
    LABELS_PAD_ID = -100
```

It wasn't obvious where this was coming from later on when you do `gt_rewards == -1000`